src/AzurePipelineParser/Models/YamlStreamProviders/ZipStreamProvider.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.IO.Compression;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace AzurePipelineParser.Models.YamlStreamProviders
+{
+    public class ZipStreamProvider : IYamlStreamProvider, IAsyncDisposable
+    {
+        private Stream? _zipStream;
+        private Lazy<ConcurrentDictionary<string, Stream>> _zipStreamData;
+        private ZipArchive _zipArchive;
+
+        public ZipStreamProvider(Stream zipStream)
+        {
+            _zipStream = zipStream;
+            _zipArchive = new ZipArchive(_zipStream, ZipArchiveMode.Read, leaveOpen: true);
+            _zipStreamData = new Lazy<ConcurrentDictionary<string, Stream>>(InitZipStreamData);
+        }
+
+        public async Task<Stream> GetStreamFromPath(string path, CancellationToken cancellationToken = default)
+        {
+            ConcurrentDictionary<string, Stream> dictionary = _zipStreamData.Value;
+
+            return await Task.Run(() =>
```

Comment:
If we are going to make all of the implementation of grabbing the stream from the dictionary synchronous anyway and await the Task.Run call, you might as well just make this all synchronous and return a Task.FromResult. I basically don't see the point of spinning up a new thread for this.